### PR TITLE
Adding collision detection based on path.

### PR DESCRIPTION
**Pull request by TravisForkner**
tick percentage will eventually be stored in a queue along with the physics objects that collided.
detect collision return 0 - 1 or null on no collision.

### CODE REVIEW

File: MitoMathHelper.js [hunk 1]
```diff
@@ -4,4 +4,82 @@
...
+        let radiusTotal = circleA.getRadius() + circleB.getRadius();
+        if(radiusTotal >= distance){
+            let backOff = Math.sqrt((radiusTotal*radiusTotal) - (distance*distance));
+            let backPoint = [backOff *( combineVelocity[0] / combineMagnitude) , backOff *( combineVelocity[1] / combineMagnitude)];
+            let closestPointWithinVelocity = this.addPoints(closestPoint , this.invertPoint(backPoint));
+            let distanceBetween = this.distanceBetweenTwoPoints(circleA.getPosition(), closestPointWithinVelocity);
+            let tickPercentage = distanceBetween / combineMagnitude;
+            console.assert(tickPercentage >= 0 && tickPercentage <= 1, "detectCollisionTime: value outside of range: " + tickPercentage );
+            return tickPercentage;
+        }else{
```
Comment (thechungster) [inline MitoMathHelper.js]: redundant else

File: MitoMathHelper.js [hunk 2]
```diff
@@ -4,4 +4,82 @@
 
 const MitoMathHelper = class MitoMathHelper {
 
+
+
+    detectCollisionTime(circleA, velocityA, circleB, velocityB, interval){
+
```
Comment (kujukuju) [inline MitoMathHelper.js]: no extra line

File: MitoMathHelper.js [hunk 3]
```diff
@@ -4,4 +4,82 @@
...
+            let tickPercentage = distanceBetween / combineMagnitude;
+            console.assert(tickPercentage >= 0 && tickPercentage <= 1, "detectCollisionTime: value outside of range: " + tickPercentage );
+            return tickPercentage;
+        }else{
+            return null;
+        }
+
+    }
+
+    distanceBetweenTwoPoints(pointA, pointB){
```
Comment (kujukuju) [inline MitoMathHelper.js]: static _distanceBetweenTwoPoints(pointA, pointB) {
include space between ) {

NOTE: you should probably return the squared hypot so you can avoid using the hypot function which calls sqrt, then only call sqrt if necessary or create another function that calls Math.sqrt(thisFunction(a, b));

File: MitoMathHelper.js [hunk 4]
```diff
@@ -4,4 +4,82 @@
...
+        return Math.hypot(pointB[0] - pointA[0], pointB[1] - pointA[1]);
+    }
+
+    detectCollisionCirCir(circleA, circleB){
```
Comment (kujukuju) [inline MitoMathHelper.js]: _detectCollisionCirleCirle(circleA, circleB) {

I wouldnt use abbreviations unless its really common

File: MitoMathHelper.js [hunk 5]
```diff
@@ -4,4 +4,82 @@
...
+
+
+    detectCollisionTime(circleA, velocityA, circleB, velocityB, interval){
+
+        let velocityIntA = this.multiplyPoint(velocityA, interval);
+        let velocityIntB = this.multiplyPoint(velocityB, interval);
+        let combineVelocity = this.addPoints(velocityIntA, this.invertPoint(velocityIntB));
+        let combineMagnitude = Math.hypot(combineVelocity[0], combineVelocity[1]);
+        //Early escape: Calculate bounds based on combine velocities
+        let bounds = new MitoCircle();
```
Comment (kujukuju) [inline MitoMathHelper.js]: you can just use a value and just do math on numbers

File: MitoMathHelper.js [hunk 6]
```diff
@@ -4,4 +4,82 @@
...
+        bounds.setPosition(circleA.getPosition()[0], circleA.getPosition()[1]);
+        bounds.setRadius(circleA.getRadius() +  combineMagnitude);
```
Comment (kujukuju) [inline MitoMathHelper.js]: no double space

File: MitoMathHelper.js [hunk 7]
```diff
@@ -4,4 +4,82 @@
...
+        if(!this.detectCollisionCirCir(bounds, circleB)){
```
Comment (kujukuju) [inline MitoMathHelper.js]: if (!MitoMathHelper._detectCollisionCircleCircle(bounds, circleB)) {

File: MitoMathHelper.js [hunk 8]
```diff
@@ -4,4 +4,82 @@
 
 const MitoMathHelper = class MitoMathHelper {
 
+
```
Comment (kujukuju) [inline MitoMathHelper.js]: remove extra lines here

File: MitoMathHelper.js [hunk 9]
```diff
@@ -4,4 +4,82 @@
...
+    invertPoint(point){
+        return [point[0] * -1 , point[1] * -1];
+    }
+
+    multiplyPoint(point, value){
+        return [point[0] * value , point[1] * value];
+    }
+
+
```
Comment (kujukuju) [inline MitoMathHelper.js]: at the end of the file do
MitoMathHelper.EPSILON = 0.000001;

when determining if things are perfectly equal use "Math.abs(value1 - EQUAL_VALUE) < MitoMathHelper.EPSILON" when applicable

File: MitoMathHelper.js [hunk 10]
```diff
@@ -4,4 +4,82 @@
 
 const MitoMathHelper = class MitoMathHelper {
 
+
+
+    detectCollisionTime(circleA, velocityA, circleB, velocityB, interval){
```
Comment (kujukuju) [inline MitoMathHelper.js]: static detectCollisionTime(circleA, velocityA, circleB, velocityB, interval) {

File: MitoMathHelper.js [hunk 11]
```diff
@@ -4,4 +4,82 @@
...
+
+        let velocityIntA = this.multiplyPoint(velocityA, interval);
```
Comment (kujukuju) [inline MitoMathHelper.js]: MitoMathHelper._multiplyPoint(velocityA, interval);

File: MitoMathHelper.js [hunk 12]
```diff
@@ -4,4 +4,82 @@
...
+        let combineMagnitude = Math.hypot(combineVelocity[0], combineVelocity[1]);
+        //Early escape: Calculate bounds based on combine velocities
+        let bounds = new MitoCircle();
+        bounds.setPosition(circleA.getPosition()[0], circleA.getPosition()[1]);
+        bounds.setRadius(circleA.getRadius() +  combineMagnitude);
+        if(!this.detectCollisionCirCir(bounds, circleB)){
+            return null;
+        }
+        //Calculate closest point on combined velocities to circleB
+        let appliedVelocitiesToPositionA = this.addPoints(circleA.getPosition(), combineVelocity);
```
Comment (kujukuju) [inline MitoMathHelper.js]: MitoMathHelper._addPoints(

File: MitoMathHelper.js [hunk 13]
```diff
@@ -4,4 +4,82 @@
...
+        let closestPoint = this.getClosestPoint(circleA.getPosition(),
+            appliedVelocitiesToPositionA,
+            circleB.getPosition());
```
Comment (kujukuju) [inline MitoMathHelper.js]: circleB.getPosition(),
);

If it lets you, it depends on the language level were using.

File: MitoMathHelper.js [hunk 14]
```diff
@@ -4,4 +4,82 @@
...
+        let closestCircle = new MitoCircle();
+        closestCircle.setPosition(closestPoint[0], closestPoint[1]);
+        closestCircle.setRadius(circleA.getRadius());
+        //If closest circle is in range calculate back off distance
+        //Tick percentage should be between 0 - 1
+        let distance = this.distanceBetweenTwoPoints(closestCircle.getPosition(), circleB.getPosition());
```
Comment (kujukuju) [inline MitoMathHelper.js]: MitoMathHelper._distanceBe...

File: MitoMathHelper.js [hunk 15]
```diff
@@ -4,4 +4,82 @@
...
+        let radiusTotal = circleA.getRadius() + circleB.getRadius();
+        if(radiusTotal >= distance){
+            let backOff = Math.sqrt((radiusTotal*radiusTotal) - (distance*distance));
```
Comment (kujukuju) [inline MitoMathHelper.js]: Math.sqrt((radiusTotal * radiusTotal) - (distance * distance));

File: MitoMathHelper.js [hunk 16]
```diff
@@ -4,4 +4,82 @@
...
+            let backPoint = [backOff *( combineVelocity[0] / combineMagnitude) , backOff *( combineVelocity[1] / combineMagnitude)];
```
Comment (kujukuju) [inline MitoMathHelper.js]: let backPoint = [backOff * (combineVelocity[0] / combineMagnitude), backOff * (combineVelocity[1] / combineMagnitude)];

File: MitoMathHelper.js [hunk 17]
```diff
@@ -4,4 +4,82 @@
...
+            let closestPointWithinVelocity = this.addPoints(closestPoint , this.invertPoint(backPoint));
```
Comment (kujukuju) [inline MitoMathHelper.js]: let closestPointWithinVelocity = this.addPoints(closestPoint, this.invertPoint(backPoint));

File: MitoMathHelper.js [hunk 18]
```diff
@@ -4,4 +4,82 @@
...
+            let distanceBetween = this.distanceBetweenTwoPoints(circleA.getPosition(), closestPointWithinVelocity);
+            let tickPercentage = distanceBetween / combineMagnitude;
+            console.assert(tickPercentage >= 0 && tickPercentage <= 1, "detectCollisionTime: value outside of range: " + tickPercentage );
```
Comment (kujukuju) [inline MitoMathHelper.js]: I wouldnt use assert's/invariant's like this, and apparently console.assert is different in node vs browsers

if you want you can be like

if (something) {
    console.error('Blah');
}

or if you are really angry

if (something) {
    throw new Error('blah');
}

File: MitoMathHelper.js [hunk 19]
```diff
@@ -4,4 +4,82 @@
...
+            return tickPercentage;
+        }else{
+            return null;
+        }
+
```
Comment (kujukuju) [inline MitoMathHelper.js]: no newline

File: MitoMathHelper.js [hunk 20]
```diff
@@ -4,4 +4,82 @@
...
+    distanceBetweenTwoPoints(pointA, pointB){
+        return Math.hypot(pointB[0] - pointA[0], pointB[1] - pointA[1]);
+    }
+
+    detectCollisionCirCir(circleA, circleB){
+        let distance = Math.hypot(
+            circleB.getPosition()[0] - circleA.getPosition()[0],
+            circleB.getPosition()[1] - circleA.getPosition()[1]);
+        let radiusTotal = circleA.getRadius() + circleB.getRadius();
+        return  radiusTotal >= distance;
```
Comment (kujukuju) [inline MitoMathHelper.js]: you can avoid using sqrt here, one of probably many places I didnt see.

no double space.

File: MitoMathHelper.js [hunk 21]
```diff
@@ -4,4 +4,82 @@
...
+    }
+
+    getClosestPoint(linePointA, linePointB, singlePoint){
+        let A_to_P = [singlePoint[0] - linePointA[0], singlePoint[1] - linePointA[1]] ;
```
Comment (kujukuju) [inline MitoMathHelper.js]: aToP, no space before ";"

File: MitoMathHelper.js [hunk 22]
```diff
@@ -4,4 +4,82 @@
...
+        let A_to_B = [linePointB[0] - linePointA[0], linePointB[1] - linePointA[1]];
```
Comment (kujukuju) [inline MitoMathHelper.js]: same

File: MitoMathHelper.js [hunk 23]
```diff
@@ -4,4 +4,82 @@
...
+        let atb2 = (A_to_B[0] * A_to_B[0]) + (A_to_B[1] * A_to_B[1]);
+
+        if(atb2 === 0){return linePointA;}
```
Comment (kujukuju) [inline MitoMathHelper.js]: if (atb2 === 0) {
    return linePointA;
}

File: MitoMathHelper.js [hunk 24]
```diff
@@ -4,4 +4,82 @@
...
+
+        let atp_dot_atb = (A_to_P[0]*A_to_B[0]) + (A_to_P[1] * A_to_B[1]);
```
Comment (kujukuju) [inline MitoMathHelper.js]: no underscores, spaces before and after *, so its consistent

File: MitoMathHelper.js [hunk 25]
```diff
@@ -4,4 +4,82 @@
...
+        let normalizedDistance = atp_dot_atb / atb2;
+        return [linePointA[0] + (A_to_B[0] * normalizedDistance) , linePointA[1] + (A_to_B[1] * normalizedDistance)]
+    }
+
+    addPoints(pointA, pointB){
+        return [pointA[0] + pointB[0] , pointA[1] + pointB[1]];
```
Comment (kujukuju) [inline MitoMathHelper.js]: no space before ","

File: MitoMathHelper.js [hunk 26]
```diff
@@ -4,4 +4,82 @@
...
+    }
+
+    invertPoint(point){
+        return [point[0] * -1 , point[1] * -1];
```
Comment (kujukuju) [inline MitoMathHelper.js]: you can do it this way, "* -1", or you can do "-point[0]"

either way is fine, I like -point[0], but you could argue that * -1 is more explicit

no space before comma

File: MitoMathHelper.js [hunk 27]
```diff
@@ -4,4 +4,82 @@
...
+    }
+
+    multiplyPoint(point, value){
+        return [point[0] * value , point[1] * value];
```
Comment (kujukuju) [inline MitoMathHelper.js]: no space before comma, private function, static function, etc

**Review by kujukuju — CHANGES_REQUESTED** [review]

File: MitoMathHelper.js [hunk 28]
```diff
@@ -4,82 +4,143 @@
...
+     * @param circleA
+     * @param velocityA
+     * @param circleB
+     * @param velocityB
+     * @param interval - tick interval
+     * @returns {*}
+     * @private
+     */
+    static _detectCollisionTime(circleA, velocityA, circleB, velocityB, interval) {
+        let velocityIntA = this._multiplyPoint(velocityA, interval);
```
Comment (kujukuju) [inline MitoMathHelper.js]: dont do this. since its static

File: MitoMathHelper.js [hunk 29]
```diff
@@ -4,82 +4,143 @@
...
+        let invertedVelocityB = this._invertPoint(this._multiplyPoint(velocityB, interval));
+        let combineVelocity = this._addPoints(velocityIntA, invertedVelocityB);
         let combineMagnitude = Math.hypot(combineVelocity[0], combineVelocity[1]);
         //Early escape: Calculate bounds based on combine velocities
-        let bounds = new MitoCircle();
-        bounds.setPosition(circleA.getPosition()[0], circleA.getPosition()[1]);
-        bounds.setRadius(circleA.getRadius() +  combineMagnitude);
-        if(!this.detectCollisionCirCir(bounds, circleB)){
+        if (!this._detectCollision(circleA.getPosition(),
```
Comment (kujukuju) [inline MitoMathHelper.js]: put the condition into a variable before the if statement since its so long

File: MitoMathHelper.js [hunk 30]
```diff
@@ -4,82 +4,143 @@
...
+            circleA.getRadius() + combineMagnitude,
+            circleB.getPosition(),
+            circleB.getRadius())) {
             return null;
         }
         //Calculate closest point on combined velocities to circleB
-        let appliedVelocitiesToPositionA = this.addPoints(circleA.getPosition(), combineVelocity);
-        let closestPoint = this.getClosestPoint(circleA.getPosition(),
+        let appliedVelocitiesToPositionA = this._addPoints(circleA.getPosition(), combineVelocity);
+        let closestPoint = this._getClosestPoint(circleA.getPosition(),
```
Comment (kujukuju) [inline MitoMathHelper.js]: move circleA.getPosition to the next line

File: MitoMathHelper.js [hunk 31]
```diff
@@ -4,82 +4,143 @@
...
+     * @private
+     */
+    static _detectCollision(pointA, radiusA, pointB, radiusB) {
         let distance = Math.hypot(
-            circleB.getPosition()[0] - circleA.getPosition()[0],
-            circleB.getPosition()[1] - circleA.getPosition()[1]);
-        let radiusTotal = circleA.getRadius() + circleB.getRadius();
-        return  radiusTotal >= distance;
+            pointA[0] - pointB[0],
+            pointA[1] - pointB[1]);
```
Comment (kujukuju) [inline MitoMathHelper.js]: pointA[1] - pointB[1],
);

you don't need to do sqrt here, find all places where you don't need to do sqrt

File: MitoMathHelper.js [hunk 32]
```diff
@@ -4,82 +4,143 @@
...
+     * @private
+     */
+    static _getClosestPoint(linePointA, linePointB, singlePoint) {
+        let aToP = [singlePoint[0] - linePointA[0], singlePoint[1] - linePointA[1]];
+        let aToB = [linePointB[0] - linePointA[0], linePointB[1] - linePointA[1]];
+        let aToB2 = (aToB[0] * aToB[0]) + (aToB[1] * aToB[1]);
+        if (aToB2 === 0) {
+            return linePointA;
+        }
+        let atpDotAtb = (aToP[0] * aToB[0]) + (aToP[1] * aToB[1]);
```
Comment (kujukuju) [inline MitoMathHelper.js]: let aToPDotAToB = ...
more descriptive naming

File: MitoMathHelper.js [hunk 33]
```diff
@@ -4,82 +4,143 @@
...
+        let normalizedDistance = atpDotAtb / aToB2;
+        return [linePointA[0] + (aToB[0] * normalizedDistance), linePointA[1] + (aToB[1] * normalizedDistance)]
```
Comment (kujukuju) [inline MitoMathHelper.js]: you dont need () around the multiplication here but it doesnt really matter either way

**Review by kujukuju — APPROVED** [review]
make changes before merging

**Pull request merged**